Added the end of booking feature along with Controller class

TeoFurcovici opened the pull request.

Commit:
Created Scooter model with additional util classes

birleadrian committed:
- created Booking model

TeoFurcovici committed:
Update SecurityConfig.java

birleadrian committed:
- removed Lombok dependencies

vladmsn committed:
Created User Model

Commit:
Created a simple scooter controller and service class skeleton

Commit:
removed lombok as dependency with all annotations from scooter model

Commit:
added nonnull check for scotter atributes

Commit:
added no args constructor and removed id

Commit:
scooter model implements serializable

Commit:
added abstract model, for storing and generating random uuid for each model

Commit:
removed abstraction for models and rewrited tostring, removed constructor

Commit:
removed uuid function call from createScooter service

TeoFurcovici committed:
Merge remote-tracking branch 'origin/birle-work' into TeoFurcovici-EndOfBooking

TeoFurcovici committed:
Merge remote-tracking branch 'origin/vladmsn-userModel' into TeoFurcovici-EndOfBooking

TeoFurcovici committed:
Merge remote-tracking branch 'origin/feature/bernat-scooter-#1' into TeoFurcovici-EndOfBooking

TeoFurcovici committed:
Merge remote-tracking branch 'origin/feature/bernat-add-scooter-#18' into TeoFurcovici-EndOfBooking

Commit:
disabled csrf, restructured firebasecontroller

birleadrian committed:
- updated Booking model

Commit:
removed unused constructors, updated equals mthods

loredanaiamnitchi committed:
Merge branch 'vladmsn-userModel' of github.com:nymphadora-tonks-team/booking-jb into view_available_scooters

loredanaiamnitchi committed:
Merge branch 'feature/bernat-scooter-#1' of github.com:nymphadora-tonks-team/booking-jb into view_available_scooters

Commit:
PR changes + created a scooter path, for the Scooter service

loredanaiamnitchi committed:
Merge branch 'feature/bernat-add-scooter-#18' of github.com:nymphadora-tonks-team/booking-jb into view_available_scooters

loredanaiamnitchi committed:
view available scooters - userLocation hardcoded

Commit:
fixed equals method for possible nullpointerexceptions

Commit:
implemented getScooterById + code cleanup

mahmoudsebaie committed:
Merge remote-tracking branch 'origin/vladmsn-userModel' into maria/booking-history

mahmoudsebaie committed:
Merge remote-tracking branch 'origin/feature/bernat-scooter-#1' into maria/booking-history

mariatirlea committed:
created service class and controller methods for viewing/ adding bookings

TeoFurcovici committed:
Adds the end of booking feature along with the Controller class

TeoFurcovici committed:
Test commit

vladmsn committed:
Updated User Model + code cleanup

Updated User Model + code cleanup

dbernatt committed:
removed unused imports

dbernatt committed:
fed csrf problem in securityconfig

dbernatt committed:
removed lombok dependency, added slf4j.api dep.

dbernatt committed:
cleanup code

TeoFurcovici committed:
Update end of booking feature with firebase

dbernatt committed:
removed lombok config file + code cleanup

dbernatt committed:
testing the new batterystatus in the battery component + code cleanup

vladmsn committed:
Overload constructor added and NotNull parameter check

dbernatt committed:
new scooter battery status solution based on the level value

vladmsn committed:
Update User.java

dbernatt commented on src/main/java/com/bootcamp/demo/security/SecurityConfig.java:
```diff
@@ -26,6 +26,7 @@ public void configure(final HttpSecurity http) throws Exception {
                 .and()
                 .logout()
                 .permitAll();
+    http.csrf().disable();
```
we no loner use csrf.disable()

dbernatt commented on src/main/java/com/bootcamp/demo/service/BookingService.java:
```diff
@@ -0,0 +1,24 @@
+    private static final Logger LOGGER = Logger.getLogger(BookingService.class.getName());
+    private static final String COLLECTION_PATH = "bookings";
+
+    private Firestore db;
```
should be: final Firestore db + you are missing the assgination in the constructor. Without the constructor you never be able to access the db.

dbernatt commented on src/main/java/com/bootcamp/demo/service/BookingService.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.logging.Logger;
+
+public class BookingService implements  IBookingService{
+    private static final Logger LOGGER = Logger.getLogger(BookingService.class.getName());
```
We use the `Slf4j` logger. Please update it accordingly. Checkout in other PRs. Ypu need to add the `org.slf4j.api `dependency

vladmsn committed:
Update User.java

TeoFurcovici committed:
Merge remote-tracking branch 'origin/maria/bookings-history' into TeoFurcovici-EndOfBooking

# Conflicts:
#	src/main/java/com/bootcamp/demo/FirebaseController.java
#	src/main/java/com/bootcamp/demo/model/util/Location.java
#	src/main/java/com/bootcamp/demo/service/BookingService.java
#	src/main/java/com/bootcamp/demo/service/IBookingService.java

TeoFurcovici committed:
Merge remote-tracking branch 'origin/view_available_scooters' into TeoFurcovici-EndOfBooking

# Conflicts:
#	src/main/java/com/bootcamp/demo/firebase/FirebaseController.java
#	src/main/java/com/bootcamp/demo/restapi/ScooterController.java
#	src/main/java/com/bootcamp/demo/security/SecurityConfig.java
#	src/main/java/com/bootcamp/demo/service/ScooterServiceImpl.java

TeoFurcovici committed:
Merge remote-tracking branch 'origin/feature/bernat-add-scooter-#18' into TeoFurcovici-EndOfBooking

# Conflicts:
#	src/main/java/com/bootcamp/demo/model/util/Battery.java
#	src/main/java/com/bootcamp/demo/restapi/ScooterController.java
#	src/main/java/com/bootcamp/demo/security/SecurityConfig.java
#	src/main/java/com/bootcamp/demo/service/ScooterService.java
#	src/main/java/com/bootcamp/demo/service/ScooterServiceImpl.java

TeoFurcovici committed:
Merge remote-tracking branch 'origin/master' into TeoFurcovici-EndOfBooking

# Conflicts:
#	src/main/java/com/bootcamp/demo/security/SecurityConfig.java
#	src/main/resources/application.properties

TeoFurcovici committed:
Merge remote-tracking branch 'origin/vladmsn-userModel' into TeoFurcovici-EndOfBooking

# Conflicts:
#	src/main/java/com/bootcamp/demo/model/Booking.java
#	src/main/java/com/bootcamp/demo/model/PaymentStatus.java

TeoFurcovici committed:
Deleted all to make the code work again

TeoFurcovici committed:
Brought back the code

TeoFurcovici committed:
Added Scooter update functionality

nymphadora-tonks-team reviewed (commented):
Usually, it is best to add a description for the pull request, in order to be easy for reviewers to see the purpose and what is intended to be changed.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+
+
+/**
+ * Sample RestController
```
You should either update this in order to fit the class or remove it. It is best to ensure that the javadoc is in sync with the implementation to not cause confusion.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+
+    @PostMapping
+    @ResponseStatus(value = HttpStatus.CREATED)
+    public ResponseEntity<Object> createBooking(@RequestBody final Booking booking) {
```
Since some functionality is common from this controller with what @mariatirlea did in https://github.com/nymphadora-tonks-team/booking-jb/pull/22, you should sync and see which version is needed.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+            LOGGER.info("Booking created successfully. Update time: " + bookingService.createBooking(booking));
+            return SUCCESS_RESPONSE;
+        } catch (ExecutionException | InterruptedException | IllegalArgumentException e) {
+            e.printStackTrace();
```
You could use logger for logging exceptions as well. If you did that, all details will be logged using the same format and in the same place.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    @PutMapping("/delete")
```
Could you highlight why you have used put and not delete/post? Also, should this method be annotated with `@ResponseBody`?

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    public double endOfBooking(Booking booking, Scooter scooter) throws ExecutionException, InterruptedException {
```
Will this be called as part of `updateBooking` or of another method? If in `updateBooking`, perhaps a more intuitive name should be used (and not one that is so general).
Also, you should renamed this method to better highlight what it does.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+    public double endOfBooking(Booking booking, Scooter scooter) throws ExecutionException, InterruptedException {
+        long bookingDuration=0;
+        double totalCostComputed=0.0;
+        if(booking.getEndDate().isAfter(booking.getEndDate())) {
```
Is this condition ok?

TeoFurcovici replied:
here, we disscued in the last meeting, it was my fault due to a typo, but now I correct it. :)

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+        long bookingDuration=0;
+        double totalCostComputed=0.0;
+        if(booking.getEndDate().isAfter(booking.getEndDate())) {
+            bookingDuration = (booking.getEndDate().getMinute() - booking.getStartDate().getMinute());
```
No need for the enclosing ( ).

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+            bookingDuration = (booking.getEndDate().getMinute() - booking.getStartDate().getMinute());
+
+        }
+        totalCostComputed=booking.getTotalCost(bookingDuration);
```
You could both declare & initialize this here, e.g. `final double totalCostComputed`

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+
+        }
+        totalCostComputed=booking.getTotalCost(bookingDuration);
+        if(totalCostComputed>0) {
```
If no payment is needed, shouldn't the booking (e.g. to mark that payment is not needed) & scooter (e.g. to mark as available) be updated as well?

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/BookingsController.java:
```diff
@@ -0,0 +1,135 @@
+            scooter.setStatus(ScooterStatus.AVAILABLE);
+            booking.setPayment(PaymentStatus.SUCCESS);
+            bookingService.updateBooking(booking.getId(), booking.getEndDate(), booking.getPayment());
+            scooter.setCurrentLocation(new Location(scooter.getCurrentLocation().getLongitude(),scooter.getCurrentLocation().getLongitude()));
```
Shouldn't update be called also for scooter?

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/restapi/ScooterController.java:
```diff
@@ -0,0 +1,67 @@
+        scooterService.createScooter(scooter);
+    }
+    @PutMapping("/updateScooter")
+    public ResponseEntity<Object> updateScooter (final String scooterId, Location location, ScooterStatus newStatus, Battery newBattery)
```
You could add `final` for all parameters for consistency.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/service/BookingService.java:
```diff
@@ -0,0 +1,115 @@
+                .whereEqualTo("accountId", userId).get().get()
+                .getDocuments()
+                .stream().map(d -> d.toObject(Booking.class))
+                .collect(Collectors.toSet())
+                .stream().sorted(Comparator.comparing(Booking :: getStartDate))
```
There is no need to call stream.collect.stream (with the intermediate collect), as it will be the same as the following:
```
.stream()
.map(d -> d.toObject(Booking.class))
.sorted(Comparator.comparing(Booking :: getStartDate))
```
Same applies for `getAllBookings`.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/service/BookingService.java:
```diff
@@ -0,0 +1,115 @@
+    @Override
+    public String deleteBooking(final UUID bookingId) throws ExecutionException, InterruptedException {
+        if (bookingId == null) throw new IllegalArgumentException();
+        ApiFuture<WriteResult> collectionApiFuture = db.collection(COLLECTION_PATH)
```
You could chain all calls, no need to use intermediate variable.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/service/BookingService.java:
```diff
@@ -0,0 +1,115 @@
+    public LinkedHashSet<Booking> getBookings(final UUID userId) throws ExecutionException, InterruptedException, IllegalArgumentException {
+        if (userId == null) throw new IllegalArgumentException();
+        LinkedHashSet<Booking> userBookings;
+        userBookings = db.collection(COLLECTION_PATH)
```
You can return the result of the call directly, without using a variable.

nymphadora-tonks-team commented on src/main/java/com/bootcamp/demo/service/ScooterServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public String updateScooter(final String scooterId, Location location, ScooterStatus newStatus, Battery newBattery) throws ExecutionException, InterruptedException {
+        LOGGER.info("CREATE SCOOTER - service function invoked:");
+        DocumentReference docRef = db.collection(COLLECTION_SCOOTERS_PATH)
```
Same here as well, you could chain all calls and return the result, without using variables.

TeoFurcovici committed:
updated the updateScooter method

TeoFurcovici committed:
another commit for UpdateScooter

dbernatt committed:
Auto stash before merge of "TeoFurcovici-EndOfBooking" and "origin/TeoFurcovici-EndOfBooking"

dbernatt committed:
code cleanup update -> set in updateScooter

TeoFurcovici committed:
code clean up and improvements

Made the neccessary improvement, so it all the operations on booking works (this was done with Maria). Beside this, based on the tips you provide me with, I made the modifications.

TeoFurcovici committed:
issue solved
